src/fu-engine.c

```diff
@@ -1638,6 +1604,82 @@ fu_engine_install (FuEngine *self,
 					 error);
 }
 
+/**
+ * fu_engine_install_store:
+ * @self: A #FuEngine
+ * @device_id: A device ID
+ * @store: The #AsStore with the firmware metadata
+ * @blob_cab: The #GBytes of the .cab file
+ * @flags: The #FwupdInstallFlags, e.g. %FWUPD_DEVICE_FLAG_UPDATABLE
+ * @error: A #GError, or %NULL
+ *
+ * Installs a specfic firmware file on a device.
+ *
+ * Returns: %TRUE for success
+ **/
+gboolean
+fu_engine_install_store (FuEngine *self,
+			 const gchar *device_id,
+			 AsStore *store,
+			 GBytes *blob_cab,
+			 FwupdInstallFlags flags,
+			 GError **error)
+{
+	FuDevice *device;
+	g_autoptr(GPtrArray) devices = NULL;
+	g_autoptr(GString) str = g_string_new (NULL);
+	gboolean overall = TRUE;
+
+	/* find the specific device */
+	if (g_strcmp0 (device_id, FWUPD_DEVICE_ID_ANY) != 0) {
+		device = fu_device_list_get_by_id (self->device_list, device_id, error);
+		if (device == NULL)
+			return FALSE;
+		return fu_engine_install (self, device, store, blob_cab, flags,
+					  error);
+	}
+
+	devices = fu_device_list_get_all (self->device_list);
+	for (guint i = 0; i < devices->len; i++) {
+		g_autoptr(GError) error_local = NULL;
```

Comment:
This can be moved closer to use, unless you continue on no GUID.

Reply:
I continued on no GUID now.